### PR TITLE
Add acceptance test specifications for all features

### PR DESCRIPTION
**Pull request by MarioBanovac**

### CODE REVIEW

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    if (state.classRoom?.id) {
+      return await prisma.assignment.create({
+        data: {
+            classId: state.classRoom.id,
+            title: state.title
+        }
+    }); 
+    }
+  
+  return undefined
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentBuilder.ts:39]: It's just my personal preference, but I'd let my builder to throw a Prisma error if classroom id is not defined. The tests would fail either way, but the return type of the function would be always an Assignment which would make it simpler IMO.

That way you could avoid optional chaining/undefined checks in your tests.

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentSubmissionBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import { StudentAssignment } from '@prisma/client';
+import { prisma } from '../../src/database';
+
+interface IAssignmentSubmissionBuilder {
+  fromAssignment: (assignment: StudentAssignment | undefined) => IAssignmentSubmissionBuilder;
+  build: () => Promise<StudentAssignment | undefined>;
+}
+
+export default function assignmentSubmissionBuilder(): IAssignmentSubmissionBuilder {
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentSubmissionBuilder.ts:9]: Nice, I think it's a great idea to create a builder like this. It's a lot more expressive than simply using a StudentAssignment builder since it expresses the submission. Love it! ❤️

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentSubmissionBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+              assignmentId: state.assignment?.assignmentId,
+              studentId: state.assignment?.studentId
+          }
+      },
+      data: {
+          grade: 'A',
+      }
+  });
+    }
+    return undefined;
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentSubmissionBuilder.ts:48]: just being nitpicky again, I'd prefer a single return type for the builder and let it throw a prisma error if something goes wrong.

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/studentAssignmentBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  const build = async () => {
+    if(state.student?.id && state.assignment?.id) {
+      return await prisma.studentAssignment.create({
+        data: {
+            studentId: state.student?.id,
+            assignmentId: state.assignment?.id,
+        }
+    });
+    }
+  return undefined
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/studentAssignmentBuilder.ts:38]: just being nitpicky again, I'd prefer a single return type for the builder and let it throw a prisma error if something goes wrong.

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/enroll_student_to_a_class.test.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    });
+
+    then("the system should return an error indicating the class is not found", () => {
+      expect(response.status).toBe(404);
+      expect(response.body.success).toBeFalsy();
+      expect(response.body.error).toBe("ClassNotFound");
+    });
+  });
+
+  test("Fail to enroll a student already enrolled in a class", ({ given, when, then }) => {
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/enroll_student_to_a_class.test.ts:84]: nice edge-case!

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_a_student_by_id.test.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    });
+
+    when('I request to access the student details', async () => {
+       response = await request(app).get(`/students/${student.id}`)
+    });
+
+    then('I should recieve the student details', () => {
+       expect(response.status).toBe(200)
+       expect(response.body.data.email).toBeDefined()
+       expect(response.body.data.name).toBeDefined()
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_a_student_by_id.test.ts:47]:
```suggestion
       expect(response.body.data.email).toBe(student.email)
       expect(response.body.data.name).toBe(student.name)
```

I think it's worth checking if it's saved with the correct data :)

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_all_students.test.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+       response = await request(app).get('/students')
+    });
+
+    then('I should get all existing students', () => {
+       expect(response.body.data.length).toBe(3)
+       expect(response.body.success).toBeTruthy()
+    });
+});
+
+test('Failed get all students', ({ given, when, then }) => {
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_all_students.test.ts:49]: nice failure case :)

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_an_assignment_by_id.test.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        assignment = await assignmentBuilder().fromClassRoom(classRoom).withTitle('Winter test').build()
+    });
+
+    when('I request to get information about the assignment', async () => {
+        response = await request(app).get(`/assignments/${assignment?.id}`)
+    });
+
+    then('I should receive the information', () => {
+        expect(response.body.data.classId).toBe(classRoom.id)
+        expect(response.body.data.title).toBe('Winter test')
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_an_assignment_by_id.test.ts:39]:
```suggestion
        expect(response.body.data.title).toBe(assignment.title)
```

Extra nitpicky, but this makes the test more resilient IMO. On the other hand, I'd say it also makes it slightly less readable, and error-prone to bugs in the builder. So dealer's choice

**Review by enekesabel — APPROVED** [review]
Very neat, readable solution as always, can't complain just approve :)

Ps.: Ofc I can, left some minor comments, but it's all up to personal preference.

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentBuilder.ts (hunk 1 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentBuilder.ts:39]: I didn't think about it that way. Thanks for the tip!

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentSubmissionBuilder.ts (hunk 3 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentSubmissionBuilder.ts:48]: I agree

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/studentAssignmentBuilder.ts (hunk 4 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/studentAssignmentBuilder.ts:38]: Agree haha

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/enroll_student_to_a_class.test.ts (hunk 5 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/enroll_student_to_a_class.test.ts:84]: Thanks

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_a_student_by_id.test.ts (hunk 6 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_a_student_by_id.test.ts:47]: Of course. I think this was generated by GPT because I was tired doing it manually haha

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_all_students.test.ts (hunk 7 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_all_students.test.ts:49]: Thanks

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_an_assignment_by_id.test.ts (hunk 8 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/acceptance/get_an_assignment_by_id.test.ts:39]: Yeah, it's a two way street but your suggestion is I would say more cleaner
